docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,844 @@
+
+The operators watch the `ControllerRequest` resources and act on those which are referencing a `ControllerRegistration` they deployed earlier.
+Gardener is responsible for writing the `.spec` field, the operator is responsible for providing information in the `.status` indicating whether the controller was successfully deployed and is ready to be used.
+Gardener will wait until all `ControllerRequest` resources indicate readiness before actually starting to reconcile or delete a `Shoot`.
+
+Gardener will be also able to delete controllers from Seeds when they are not needed there anymore by deleting the corresponding `ControllerRequest` object.
+
+:information: The provided easy-to-use framework for the controllers will also contain these needed features to implement corresponding operators.
+
+### Helm Chart-based approach
+
+Every controller registers itself at the Gardener by creating a `ControllerRegistration` resource that make the controller together with its dimension (`kind`) and shape (`type`) known to Gardener.
+This registration resource does also contain a Helm chart with corresponding values:
+
+```yaml
+apiVersion: gardener.cloud/v1alpha1
+kind: ControllerRegistration
+metadata:
+  name: dns-aws-route53
+spec:
+  controllerInfo:
+  - kind: DNS
+    type: aws-route53
+  deployment:
+    chart.tgz: base64(helm-chart-blob)
+    values.yaml: base64(corresponding static chart values)
+status:
+  seeds:
+  - name: seed-01
+    ready: false
+```
+
+Every `.kind`/`.type` combination may only exist once in the system.
+
+With knowledge of the Helm Chart and the corresponding values Gardener will be able to deploy the controllers themselves in the needed Seeds.
+Gardener will track in the `.status` of the resource to which Seeds it has already which controller.
+With this approach, cleaning up is not easily/cleanly possible as Gardener would need to parse the Helm chart to understand which resources must be deleted.
+Updating or upgrading is also more difficult in case the existing deployed resources can't be "re-applied" blindly.
+Another problem could be that the `values.yaml` is static and cannot differ between Seed clusters.
+Still, the reason for mentioning this approach despite the rather longer list of downsides is that a major benefit would be to not require dedicated operators per controller.
+One could argue that in most cases static values are fine and updating is as easy as just exchanging the image/tag of the controller.
+
+## Alternatives
+
+* Alternative to DNS approach for Shoot control plane movement/migration: We have thought about rotating the credentials when a move is triggered which would make all controllers ineffective immediately. However, one problem with this is that we require IAM privileges for the users infrastructure account which might be not desired. Another, more complicated problem is that we cannot assume API access in order to create technical users for all cloud environments that might be supported.
```

Comment:
In general, I am still a bit concerned about the DNS lock/synchronisation. It is certainly a good approach, but even better would be strict credentials rotation (Gardener requests the new control plane to rotate the creds, so that the old control plane loses access and can't regain it again). However, as you pointed out, there is the problem with missing IAM privileges and some awkward/backward infrastructures that require manual interactions for new technical users, which clearly break any idea of cred rotation and security. Sigh...

Reply:
One possibility to work around the limitations of *some* infrastructures regarding interactions for new technical users would be for Gardener to require two sets of credentials.
1. Technical user actually used by Gardener to make infrastructure API calls (with just the exact rights required).
2. Technical user with just the exact rights to be able to reset/rotate the credentials for the technical user above.

From the last interaction with OpenStack, the 2 seems possible in an automated way.

* Pros
  * This would eliminate the need for full IAM access. Just for resetting/rotating credentials for a particular technical user.
  * This would work around the need to create new technical users (manually of automatically).
  * This would address security concerns regarding credentials rotation.
* Cons
  * This would complicate the initial setup.